android/src/main/java/org/ligi/gobandroid_hd/logic/GoMove.java

```diff
@@ -19,75 +19,131 @@
+                move = move.parent;
             }
         }
     }
 
     public GoMove(Cell cell, GoMove parent) {
         this(parent);
-
         this.cell = cell;
     }
 
-    public void setDidCaptures(boolean did) {
-        did_captures = did;
+    public GoMove(Cell cell, GoMove parent, StatefulGoBoard board) {
+        this(cell, parent);
+        byte previousStatus = board.getCellKind(cell);
+        board.setCell(cell, (byte) getCellStatus());
+        buildCaptures(board);
+        board.setCell(cell, previousStatus);
+    }
+
+    public boolean isIllegalKo() {
+        return parent != null &&
+            captures.size() == 1 &&
+            parent.captures.size() == 1 &&
+            parent.captures.get(0).equals(cell);
+    }
+
+    public boolean isIllegalNoLiberties(StatefulGoBoard board) {
+        if(!captures.isEmpty()) {
+            return false;
+        }
+
+        byte previousStatus = board.getCellKind(cell);
+        board.setCell(cell, (byte) getCellStatus());
+        boolean hasLiberties = board.doesCellGroupHaveLiberty(cell);
+        board.setCell(cell, previousStatus);
+        return !hasLiberties;
+    }
+
+    public void apply(StatefulGoBoard board) {
+        parent.addNextMove(this);
+        board.setCell(cell, (byte) getCellStatus());
+        board.setCellGroup(captures, STONE_NONE);
+    }
+
+    public GoMove undo(StatefulGoBoard board, boolean keepMove) {
+        if(parent == null) {
+            return this;
+        } else if(!keepMove) {
+            parent.next_move_variations.remove(this);
+        }
+
+        if(cell != null) {
+            board.setCell(cell, STONE_NONE);
+            for(Cell capture : captures) {
+                board.setCell(capture, (byte) getCapturedCellStatus());
+            }
+        }
+
+        return parent;
+    }
+
+    public GoMove redo(StatefulGoBoard board, GoMove next) {
+        if(!next_move_variations.contains(next)) {
+            return this;
+        }
+
+        next.apply(board);
+        return next;
     }
 
-    public boolean isCapturesMove() {
-        return did_captures;
+    public void buildCaptures(StatefulGoBoard board) {
+        StatelessBoardCell boardCell = board.getCell(cell);
+        for(Cell neighbor : boardCell.getNeighbors()) {
+            if(!captures.contains(neighbor) && !board.isCellFree(neighbor) && !board.areCellsEqual(neighbor, cell) && !board.doesCellGroupHaveLiberty(neighbor)) {
+                Set<StatelessBoardCell> cellGroup = board.getCellGroup(neighbor);
+                captures.addAll(cellGroup);
+            }
+        }
     }
 
     public int getMovePos() {
         return move_pos;
     }
 
+    public byte getPlayer() {
+        return player;
+    }
+
+    public void setPlayer(byte player) {
+        this.player = player;
+    }
+
     public boolean hasNextMove() {
-        return (next_move_variations.size() > 0);
+        return !next_move_variations.isEmpty();
```

Comment:
List doesn't have an isNotEmpty method (though I wish it did). Are you thinking of some utility method like SomeUtil.isNotEmpty(list)?